micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java

```diff
@@ -0,0 +1,265 @@
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.streams.KafkaStreams;
+
+import static java.util.Collections.emptyList;
+
+/**
+ * Kafka metrics binder.
+ * <p>
+ * It is based on {@code metrics()} method returning {@link Metric} map exposed by clients and
+ * streams interface.
+ * <p>
+ * Meter names have the following convention: {@code kafka.(metric_group).(metric_name)}
+ *
+ * @author Jorge Quilcate
+ * @see <a href="https://docs.confluent.io/current/kafka/monitoring.html">Kakfa monitoring
+ * documentation</a>
+ * @since 1.4.0
+ */
+@Incubating(since = "1.4.0")
+@NonNullApi
+@NonNullFields
+public class KafkaMetrics implements MeterBinder {
+    static final String METRIC_NAME_PREFIX = "kafka.";
+
+    static final String METRIC_GROUP_APP_INFO = "app-info";
+    static final String VERSION_METRIC_NAME = "version";
+    static final String START_TIME_METRIC_NAME = "start-time-ms";
+
+    private final Supplier<Map<MetricName, ? extends Metric>> metricsSupplier;
+    private final Iterable<Tag> extraTags;
+
+    /**
+     * Keeps track of current set of metrics. When this values change, metrics are bound again.
+     */
+    private volatile Map<MetricName, Meter> currentMeters = new HashMap<>();
+
+    private String version = "";
+
+    /**
+     * Kafka {@link Producer} metrics binder
+     *
+     * @param kafkaProducer producer instance to be instrumented
+     * @param tags          additional tags
+     */
+    public KafkaMetrics(Producer<?, ?> kafkaProducer, Iterable<Tag> tags) {
+        this(kafkaProducer::metrics, tags);
+    }
+
+    /**
+     * Kafka {@link Producer} metrics binder
+     *
+     * @param kafkaProducer producer instance to be instrumented
+     */
+    public KafkaMetrics(Producer<?, ?> kafkaProducer) {
+        this(kafkaProducer::metrics);
+    }
+
+    /**
+     * Kafka {@link Consumer} metrics binder
+     *
+     * @param kafkaConsumer consumer instance to be instrumented
+     * @param tags          additional tags
+     */
+    public KafkaMetrics(Consumer<?, ?> kafkaConsumer, Iterable<Tag> tags) {
+        this(kafkaConsumer::metrics, tags);
+    }
+
+    /**
+     * Kafka {@link Consumer} metrics binder
+     *
+     * @param kafkaConsumer consumer instance to be instrumented
+     */
+    public KafkaMetrics(Consumer<?, ?> kafkaConsumer) {
+        this(kafkaConsumer::metrics);
+    }
+
+    /**
+     * {@link KafkaStreams} metrics binder
+     *
+     * @param kafkaStreams instance to be instrumented
+     * @param tags         additional tags
+     */
+    public KafkaMetrics(KafkaStreams kafkaStreams, Iterable<Tag> tags) {
+        this(kafkaStreams::metrics, tags);
+    }
+
+    /**
+     * {@link KafkaStreams} metrics binder
+     *
+     * @param kafkaStreams instance to be instrumented
+     */
+    public KafkaMetrics(KafkaStreams kafkaStreams) {
```

Comment:
I think we'll have to split this out to another class. Compilation will fail when using `KafkaMetrics` without `kafka-streams` on the compilation classpath.